### PR TITLE
Execute attribute mappers on client initiated account linking (#11195)

### LINKED ISSUE #11195: Sync user attributes on identity provider account linking
### Describe the bug

Currently, user attributes do not seem to sync when using account linking from an identity provider, despite mappers being set up for the identity provider.

I can see a PR open that seems to address the issue (though no issue open on Github as far as I can see), could this be reviewed?

https://github.com/keycloak/keycloak/pull/8139

### Version

17.0.1

### Expected behavior

When a user has their account linked from an identity provider and attribute mappers are set up for the provider, the attributes are created for the Keycloak user using their identity provider account.

### Actual behavior

A user can successfully link to an identity provider account but no attributes are created for the user.

### How to Reproduce?

Enable account linking only for an identity provider and mappers for some attributes.

### Anything else?

PR seems to be open here:
https://github.com/keycloak/keycloak/pull/8139

### PR DESCRIPTION
**Pull request by toddkazakov**
Fixes #11195

### CODE REVIEW

**Comment by thomasdarimont** [comment]
Thanks for your contribution! I think this is indeed useful. Just had a small note about the imports that needs fixing.

**Review by thomasdarimont — CHANGES_REQUESTED** [review]

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/AccountLinkTest.java [hunk 1]
```diff
@@ -22,19 +22,17 @@
...
 import org.keycloak.common.Profile.Feature;
 import org.keycloak.common.util.MultivaluedHashMap;
-import org.keycloak.models.KeycloakSession;
-import org.keycloak.models.RealmModel;
-import org.keycloak.models.UserModel;
-import org.keycloak.representations.idm.ComponentRepresentation;
-import org.keycloak.representations.idm.FederatedIdentityRepresentation;
-import org.keycloak.representations.idm.RealmRepresentation;
-import org.keycloak.representations.idm.UserRepresentation;
+import org.keycloak.models.*;
```
[the remark this answers is not in the document]
Comment (toddkazakov) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/AccountLinkTest.java]: Thanks! Fixed.

**Comment by toddkazakov** [comment]
Unfortunately, I'm unable to run the test suite locally, due to the jakarta migration. For some reason maven tries to use `keycloak-admin-client-jee` instead of `keycloak-admin-client` in the integration test suite.

**Review by andre-nascimento6791 — COMMENTED** [review]
I hope to test it in my Env ASAP, but it LGTM.

Good Job, @toddkazakov .

**Comment by toddkazakov** [comment]
The added test doesn't work after rebase because of #15199. I'm working on moving it in a different spot, because the account linking mappers are not triggered with the refactored logic.

**Comment by toddkazakov** [comment]
@andre-nascimento-rh tests should be ok now

**Review by andre-nascimento6791 — COMMENTED** [review]
@toddkazakov , I didn't have time to test it locally yet. Sorry. But from the coding perspective, it LGTM.

Let's see If all tests pass in GHA.

Thanks.

File: services/src/main/java/org/keycloak/services/resources/IdentityBrokerService.java [hunk 2]
```diff
@@ -974,6 +974,8 @@ private Response performAccountLinking(AuthenticationSessionModel authSession, U
             userSession.setNote(Details.IDENTITY_PROVIDER_USERNAME, context.getUsername());
         }
 
+        updateFederatedIdentity(context, federatedUser);
```
Comment (pedroigor) [inline services/src/main/java/org/keycloak/services/resources/IdentityBrokerService.java:977]: @toddkazakov I think calling this method here will cause errors when linking existing accounts. Could you please try to add the call at https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/services/resources/IdentityBrokerService.java#L942?

**Comment by pedroigor** [comment]
@toddkazakov Sorry for the back and forth ...

I just realized that the test is not really covering the change I suggested. It is only passing because right after the first attempt to link the account, the test tries again to link the account and only at this time the attributes are set, as expected.

I created this branch with some minor updates. See https://github.com/pedroigor/keycloak/tree/issue-11195. 

I also found a difference in behavior when using legacy and new stores. This code returns null when using the new store (chm) but using the current store resolves the model instance. See https://github.com/pedroigor/keycloak/commit/e7b2041b2d0bb1231286114b555f978fa8bb1767#diff-e44efb607cb8cb833e99677c01e7fc10d67afc6ee54ce4faf06a01f5acaf50e7R953.

@keycloak/store-maintainers  Is this difference expected? Looks like changes made during a transaction are not immediately visible until it is committed.

**Comment by toddkazakov** [comment]
@pedroigor no problem, makes sense. In the tests it's probably better to unlink the account after the first test is successful. 

How do you want to proceed with this one? Seems like the new store not seeing the uncommitted changes is a no go.

**Comment by pedroigor** [comment]
@toddkazakov Thanks. I'm looking at the new store-related changes. I'll ping you back once my branch is updated.

**Comment by pedroigor** [comment]
@toddkazakov Do you mind looking at https://github.com/keycloak/keycloak/pull/20965?

**Comment by toddkazakov** [comment]
@pedroigor looks ok to me. Were there any changes related to the new store?

**Comment by pedroigor** [comment]
@toddkazakov TBH, I'm not sure. I could not reproduce the behavior anymore.

Do you mind updating your PR (sorry, again) based on https://github.com/keycloak/keycloak/pull/20965?

**Comment by pedroigor** [comment]
Fixed by https://github.com/keycloak/keycloak/pull/20965

**Pull request closed without merging**